src/open-hackathon-server/Kaiyuanshe.OpenHackathon.Server/Storage/BlobContainers/AzureBlobContainer.cs

```diff
@@ -0,0 +1,74 @@
+﻿using System;
+using Microsoft.WindowsAzure.Storage;
+using Microsoft.WindowsAzure.Storage.Blob;
+
+namespace Kaiyuanshe.OpenHackathon.Server.Storage.BlobContainers
+{
+
+    public interface IAzureBlobContainer
+    {
+        string CreateSasToken();
+    }
+
+    public class AzureBlobContainer: IAzureBlobContainer
+    {
+        #region Constants
+        /// <summary>
+        /// The retry interval between for blob container retry operations
+        /// </summary>
+        static readonly int BlobContainerRetryInterval = 5;
+
+        /// <summary>
+        /// The retry count for blob container operations
+        /// </summary>
+        static readonly int BlobContaineRetryCount = 5;
+
+        /// <summary>
+        /// The default SAS time span in minutes
+        /// </summary>
+        static readonly int BlobContainerSasTimeSpan = 2; // minutes
+        #endregion
+
+        /// <summary>
+        /// client to make request aganist azure blob container.
+        /// </summary>
+        protected readonly CloudBlobClient blobServiceProxy;
+
+        /// <summary>
+        /// proxy container that represents the azure blob container.
+        /// </summary>
+        protected readonly CloudBlobContainer blobContainerProxy;
+
+        /// <summary>
+        ///  The name of the blob container exists under the storage account with which this AzureBlobContainer is configured.
+        /// </summary>
+        protected readonly string blobName;
+
+        /// <summary>
+        /// Test-only constructor
+        /// </summary>
+        protected AzureBlobContainer()
+        { }
+
+        protected AzureBlobContainer(CloudStorageAccount storageAccount, string blobName)
+        {
+            this.blobName = blobName;
+            blobServiceProxy = storageAccount.CreateCloudBlobClient();
+            blobServiceProxy.DefaultRequestOptions.RetryPolicy = new Microsoft.WindowsAzure.Storage.RetryPolicies.LinearRetry(
+                TimeSpan.FromSeconds(BlobContainerRetryInterval), BlobContaineRetryCount);
+            blobContainerProxy = this.blobServiceProxy.GetContainerReference(this.blobName);
+        }
+
+        public string CreateSasToken()
+        {
+            var sasPolicy =
+                new SharedAccessBlobPolicy
+                {
+                    SharedAccessExpiryTime = DateTime.UtcNow.AddMinutes(BlobContainerSasTimeSpan * 2),
+                    SharedAccessStartTime = DateTime.UtcNow.Subtract(new TimeSpan(0, BlobContainerSasTimeSpan, 0)),
+                    Permissions = SharedAccessBlobPermissions.Read | SharedAccessBlobPermissions.Write
+                };
+            return blobContainerProxy.GetSharedAccessSignature(sasPolicy);
```

Comment:
The permission it too big. We cannot give the client Write access to the whole blob container. User can upload anyfile to any sub folder, which might override other user's files. 

Here are some design considerations:
- We must be grant access to one blob only. And we need to include the blob path in the response to avoid user uploading any file to any sub folder.  It's so easy to abuse.
- We also need orgainze files with proper folders to avoid all files being uploaded to the save directory. For example, including date into the path
- We may need to rename the file to avoid naming conflict(so we need to generate a random name and ask the api caller to provider the original name/format).
- We may aslo want to track some history in tables, to add the capability to enforce some restriction like blacklist, maximum file counts, maximum file size etc.

Reply:
That's indeed a huge security issue if not designed carefully.

1. To manage the scope, should we create a blob container for each individual user?
2. Any best practice that we can learn from?